account-integrations/safe/src/SafeWebAuthnPlugin.sol

```diff
@@ -67,35 +67,92 @@ contract SafeWebAuthnPlugin is BaseAccount {
     ) internal override returns (uint256 validationData) {
-        {
+        bytes calldata authenticatorData;
+        bytes calldata clientData;
+        uint256[2] calldata signature;
+        uint256[2] calldata pubKey;
+        LocalVarStruct memory s;
+
+        {            
+            // parse length of all fixed-length params (including length)
+            uint i = 0;
+            uint dataLen = 32;
+            uint256 paramLen = abi.decode(userOp.signature[i:i+dataLen], (uint256));
+            // Fixed-length params (bytes1, (uint256?), uint256, uint256[2], uint256[2]). Expect 8 slots = 256 bytes
+            i += dataLen; // advance index
+
+            // decode fixed length params (values to memory)
+            dataLen = 3 * 32; //lenFixedParams - 32; // -32 already read length
             (
-                bytes memory authenticatorData,
-                bytes1 authenticatorDataFlagMask,
-                bytes memory clientData,
-                uint256 clientChallengeDataOffset,
-                uint256[2] memory signature,
-                uint256[2] memory pubKey
+                s.authenticatorDataFlagMask,
+                , // some number
+                s.clientChallengeDataOffset
             ) = abi.decode(
-                    userOp.signature,
-                    (bytes, bytes1, bytes, uint256, uint256[2], uint256[2])
-                );
-
-            bool verified = FCL_WebAuthn.checkSignature(
-                authenticatorData,
-                authenticatorDataFlagMask,
-                clientData,
-                userOpHash,
-                clientChallengeDataOffset,
-                signature,
-                pubKey
+                userOp.signature[i:i+dataLen],
+                (
+                    bytes1,
+                    uint256, //not sure what is encoded here
+                    uint256
+                )
             );
-            if (!verified) return SIG_VALIDATION_FAILED;
-            return 0;
+            i += dataLen; // advance index
+
+
+            bytes calldata calldataLocation;
+            // load fixed length array params (pointers to calldata)
+            dataLen = 2 * 32;
+            calldataLocation = userOp.signature[i:i+dataLen];
+            assembly{
+                signature := calldataLocation.offset
+            }
+            i += dataLen; // advance index
+
+            calldataLocation = userOp.signature[i:i+dataLen];
+            assembly{
+                pubKey := calldataLocation.offset
+            }
+            i += dataLen; // advance index
+
+            // parse length of authenticatorData
+            dataLen = 32;
+            paramLen = abi.decode(userOp.signature[i:i+dataLen], (uint256));
+            i += dataLen; // advance index
+            // assign authenticatorData to sig splice
+            dataLen = paramLen;//((paramLen >> 5) + 1) << 5; // (round up to next slot)
+
+            authenticatorData = userOp.signature[i:i+dataLen];
```

Comment:
To make sure I understand how this is working:
1. `paramLen` is the length of `authenticatorData` padded to 32 bytes.
2. Once we have the length, we advance the index as we don't care about the length bytes now (with `i += dataLen;`).
3. We then slice the `authenticatorData` from the start of `authenticatorData` to `i+dataLen`, which is the total length of the `authenticatorData` bytes. Thus giving us the correct bytes?

^Please correct me if I'm wrong

One thing I'm not sure on, is why we need to assign `dataLen = paramLen;`, can you explain the reasoning behind that? I played around with the code to just use `paramLength` and the tests still passed for me:

```solidity
        // parse length of authenticatorData
        dataLen = 32;
        paramLen = abi.decode(userOp.signature[i:i + dataLen], (uint256));
        i += dataLen; // advance index
        // assign authenticatorData to sig splice
        authenticatorData = userOp.signature[i:i + paramLen];
        i += ((paramLen >> 5) + 1) << 5; // advance index (round up to next slot)

        // parse length of clientData
        paramLen = abi.decode(userOp.signature[i:i + dataLen], (uint256));
        i += dataLen; // advance index
        // assign clientData to sig splice
        clientData = userOp.signature[i:i + paramLen];
```